File: app/src/main/java/com/example/android/popularmovies/utilities/NetworkUtil.java
package com.example.android.popularmovies.utilities;

import android.net.Uri;

import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.Scanner;

public class NetworkUtil {

    private static final String URL_ADDRESS = "https://api.themoviedb.org/3/movie/";
    private static final String FORMAT = "mode";
    private static final String format = "json";
    private static final String API_KEY = "api_key";
    private static final String api_key = "3b623a17f57eb4da612b3871d3f78ced";


    public static URL buildUrl(String query){
        String urlAddress = URL_ADDRESS + query;
        Uri builtUri = Uri.parse(urlAddress).buildUpon()
                .appendQueryParameter(FORMAT, format)
                .appendQueryParameter(API_KEY, api_key)
                .build();

        URL url = null;
        try {
            url = new URL(builtUri.toString()   );
        } catch (MalformedURLException e) {
            e.printStackTrace();
        }
        return url;
    }

    public static String getResponseFromHttpUrl(URL url) throws IOException {
        HttpURLConnection urlConnection = (HttpURLConnection) url.openConnection();
        try {
            InputStream in = urlConnection.getInputStream();

            Scanner scanner = new Scanner(in);
            scanner.useDelimiter("\\A");

            boolean hasInput = scanner.hasNext();
            if (hasInput) {
                return scanner.next();
            } else {
                return null;
            }
        } finally {
            urlConnection.disconnect();
        }
    }

}

File: app/src/main/java/com/example/android/popularmovies/utilities/NetworkUtil.kt
package com.example.android.popularmovies.utilities

import android.net.Uri
import androidx.lifecycle.LiveData
import com.example.android.popularmovies.models.*
import com.squareup.moshi.Moshi
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
import retrofit2.Retrofit
import retrofit2.converter.moshi.MoshiConverterFactory
import retrofit2.http.GET
import retrofit2.http.Path
import retrofit2.http.Query
import java.io.IOException
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.URL
import java.util.*


interface TMDBService{

    @GET("/3/movie/popular")
    suspend fun getPopularMovies(@Query(API_KEY) apiKey: String = api_key): TMDBResponse

    @GET("/3/movie/top_rated")
    suspend fun getTopRatedMovies(@Query(API_KEY) apiKey: String = api_key): TMDBResponse


    @GET("/3/movie/{movieId}/videos")
    suspend fun getVideos(
            @Path("movieId") movieId: Long,
            @Query(API_KEY) apiKey: String = api_key): VideoResponse

    @GET("/3/movie/{movieId}/reviews")
    suspend fun getReviews(
            @Path("movieId") movieId: Long,
            @Query(API_KEY) apiKey: String = api_key): ReviewResponse

    @GET("/3/movie/{movieId}/recommendations")
    suspend fun getRecommendationsBaseOnMovieID(
            @Path("movieId") movieId: Long,
            @Query(API_KEY) apiKey: String = api_key): TMDBResponse

    @GET("/3/movie/{movieId}")
    suspend fun getMovieDetailsVideosReviewsById( @Path("movieId") movieId: Long,
                                                  @Query("append_to_response") appendToResponse: String = append_to_response,
                                                  @Query(API_KEY) apiKey: String = api_key): MovieDetailsResponse

}



const val IMAGE_WIDTH = 500
private const val URL_ADDRESS = "https://api.themoviedb.org"
private const val API_KEY = "api_key"
private const val api_key = "3b623a17f57eb4da612b3871d3f78ced"
const val POSTER_BASE_URI = "http://image.tmdb.org/t/p/w500"
const val VIDEO_BASE_URI = "https://www.youtube.com/watch?v="
private const val append_to_response = "videos,reviews"
private const val APPEND_TO_RESPONSE = "append_to_response"


private val moshi = Moshi.Builder().add(KotlinJsonAdapterFactory()).build()

private val retrofit = Retrofit.Builder()
        .addConverterFactory(MoshiConverterFactory.create(moshi))
        .baseUrl(URL_ADDRESS)
        .build()

object TMDBApi{
    val retrofitService : TMDBService by lazy {
        retrofit.create(TMDBService::class.java)
    }
}

